### ISSUE TITLE
Add-update-label-to-issues-weekly workflow failed from Feb 17 - Mar 8, 2022

### ISSUE THREAD

**Issue by SAUMILDHANKAR** [hackforla/website#2932]
### Overview
As a developer, I would like all our workflows to be error-free. For this issue, we will identify why the Add-update-label-to-issues-weekly workflow failed from [Feb 17 - Mar 8, 2022](https://github.com/hackforla/website/actions/workflows/add-update-label-weekly.yml), and fix the error if required.

### Action Items
- [x] Review the following files: [add-label.js](https://github.com/hackforla/website/blob/gh-pages/github-actions/add-update-label-weekly/add-label.js), [find-linked-issue.js](https://github.com/hackforla/website/blob/gh-pages/github-actions/utils/find-linked-issue.js) and [add-update-label-weekly yml file](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/add-update-label-weekly.yml)
- [x] Issues where the workflow failed: [2159](https://github.com/hackforla/website/runs/5264034478?check_suite_focus=true) and [2660](https://github.com/hackforla/website/runs/5330225436?check_suite_focus=true)
- [ ] Check if the error might be related to issue number 2634
- [ ] Identify the reasons for the error and advise on if changes need to be made to make sure this is not repeated
- [ ] Test the changes in your own repo by replicating the scenario where the workflow failed and ensure no other functionality is impacted
- [ ] Make a PR to fix the error if required.

### Resources/Instructions
- [Error log](https://github.com/hackforla/website/runs/5418754678?check_suite_focus=true)
- [add-label.js file](https://github.com/hackforla/website/blob/gh-pages/github-actions/add-update-label-weekly/add-label.js)
- [find-linked-issue.js file](https://github.com/hackforla/website/blob/gh-pages/github-actions/utils/find-linked-issue.js)
- [add-update-label-weekly yml file](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/add-update-label-weekly.yml)

**Comment by kathrynsilvaconway** [comment]
I am returning this issue to the prioritized backlog because this dev has left HfLA.

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/01/2023 Wed 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability**: Daily 10am-6pm ET (7am - 3pm PT)
**ETA**: 03/02/2023 Thu 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability**: Daily 10am-6pm ET (7am - 3pm PT)
**ETA**: 03/04/2023 Sat 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Progress**: "What is the current status of your project? What have you completed and what is left to do?"
I have inspected the files in question and the logs for the failed workflows. Failure was due to a null object. I have set up a test repository and gotten up to speed on how GitHub Actions work. I have looked at the hackforla document ["How to test GitHub Actions"](https://docs.google.com/document/d/1A2sj5KROv_AD3qTUrPO-YI730hKj2-o1FnE8GE1PUVY/edit)
What is left to do: Choose a testing methodology to see if the error can be reproduced. Decide on a course of action for possibly fixing the error if necessary.
**Blockers**: "Difficulties or errors encountered.": None
**Availability**: "How much time will you have this week to work on this issue?" Daily 10am-6pm ET (7am - 3pm PT)
**ETA**: "When do you expect this issue to be completed?" 03/06/2023 Mon 6pm ET (3pm PT)
**Pictures (optional)**: "Add any pictures of the visual changes made to the site so far." None.

**Comment by MarkWiltberger** [comment]
**Availability**: Daily 10am-6pm ET (7am - 3pm PT)
**ETA**: 03/07/2023 Tue 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability**: Daily 10am-6pm ET (7am - 3pm PT)
**ETA**: 03/08/2023 Wed 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability**: Daily 10am-6pm ET (7am - 3pm PT)
**ETA**: 03/10/2023 Fri 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/11/2023 Sat 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/12/2023 Sun 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Progress:** "What is the current status of your project? What have you completed and what is left to do?"
I have identified a condition in which there may be a null object returned by the API. I identified the condition by inspecting the objects from test issues created in my test repository returned by the API.
What is left to do: Set up a test procedure, document the testing procedure, and determine whether there is an error that needs to be fixed.
**Blockers:** "Difficulties or errors encountered.": None
**Availability:** "How much time will you have this week to work on this issue?" Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** "When do you expect this issue to be completed?" 03/13/2023 Mon 6pm ET (3pm PT)
**Pictures (optional):** "Add any pictures of the visual changes made to the site so far." None.

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/15/2023 Wed 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/17/2023 Fri 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/19/2023 Sun 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Progress:** Working on test procedures to reproduce the error, test the error, and fix the code. Need to document the procedures and fix the error, if necessary.
**Blockers:** None.
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**  03/20/2023 Mon 6pm ET (3pm PT)
**Pictures (optional):** None.

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/21/2023 Tue 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/23/2023 Thu 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/27/2023 Mon 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 03/29/2023 Wed 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Availability:** Daily 10am-6pm ET (7am - 3pm PT)
**ETA:** 04/01/2023 Sat 6pm ET (3pm PT)

**Comment by MarkWiltberger** [comment]
**Progress:** Although the log for the dates of the errors no longer exists, there is a reference to an error being generated for a method on a null property in the file find-linked-issue.js
I have set up the _jest_ testing package for node.js on my local laptop dev environment. I have reproduced the error in a simple case by passing a null property to find-linked-issue.js.
I have set up a test repository in my remote GitHub account to generate test issues. I have written code that generates test issue objects.
The next step is to write a test which passes these issue objects to find-linked-issue.js to check for pass or fail of real issues and issue conditions.
**Blockers:**
 None.
**Availability:** 
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
 04/12/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
 None.

**Comment by MarkWiltberger** [comment]
**Progress:** At the moment, the issue is difficult to test in the current state of the code because of the tight coupling, lack of separation of concerns, and general complexity.
While it is possible to test find-linked-issue.js with a simple case, it is difficult to test the code which calls it, `add-label.js`. Thus it is difficult to test `find-linked-issue.js` with real GitHub Issue data, and it is difficult to test `add-label.js` to make sure it was not broken.
I created a Static Call Graph of `add-label.js` to see the structure of the function calls, in order to analyze how the functions could be refactored so that they may be tested more easily. The diagram is embedded below under **Pictures**.
The next action is to finish analysis of the diagram and the code to determine future action item options. In addition, creating tests for calling the GitHub Action API for the Issue object is another possible action item.
**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
04/19/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
<img width="2304" alt="Figjam-hackforla-githubaction-add-label-static-call-graph 2023 04 16 mtw" src="https://user-images.githubusercontent.com/17149108/232326300-de471ffa-6509-447e-bde8-372d31c33e79.png">

**Comment by MarkWiltberger** [comment]
**Progress:** 
I have looked into how to use jest testing package to test private functions in node.js.
Next Steps are to look further into how to conduct unit tests of functions of add-label.js and find-linked-issue.js.
**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
04/26/2023 Wed 6pm ET (**3pm PT)
Pictures (optional):**

**Comment by MarkWiltberger** [comment]
**Progress:**
I have looked into how to test private functions in Node.js/Javascript.
There are several techniques, each of which is not perfect.
The most straightforward method, which avoids the installation of additional node modules or plugins and additional testing code syntax to learn, is to export the private functions in the code. The private functions are exported conditionally on an environmental variable so they will only be exported in a testing environment. Additionally, the exported object of private functions will be named with a convention that indicates to a developer that it is a private function, and not to be imported by another node module.

Next Steps are to write unit tests for the private function in the module and try the tests to see if the technique works. Additionally, the GitHub action workflow must be tested to make sure that the workflow still works correctly in the production  environment (the GitHub actions project environment).

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
05/03/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
I am in the midst of writing tests for the script functions

_Next Steps:_
Finish writing the tests and try them out.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
05/10/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
I am in the midst of writing tests for the script functions. I was able to successfully export the private functions and test them.

_Next Steps:_
Finish writing the tests and try them out.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
05/17/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
I have been writing tests for the script functions.

**Next Steps:**
Continue writing test scripts.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
05/24/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
No progress/time allotted this week on writing test scripts. Did research on testing/refactoring best practices.

_Next Steps:_
Continue writing test scripts.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
05/31/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
No progress.

_Next Steps:_
Write test scripts.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
06/07/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
Researched e2e tests in github context on local machine.

_Next Steps:_
Write test scripts.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
06/14/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
Updated the team and project leads on the issue at the Sunday DEV/All Teams meeting last Sunday, June 11.

Worked on writing integrated test script.

_Next Steps:_
Write test scripts.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
06/21/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
Worked on writing test scripts.

_Next Steps:_
Continue working on writing test scripts.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
06/28/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
Looked at github and context objects in action-script but otherwise did not have time to make much progress.

_Next Steps:_
Write test scripts that are started via Github Actions using Act/Nektos or via node.js.
I do not have much time over the next week to work on this so I am setting the ETA for 10 days out.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
07/12/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
I have been traveling and unable to make progress on this issue the past week.

_Next Steps:_
Write test scripts that are started via Github Actions using Act/Nektos or via node.js.


**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
07/26/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
I have been on vacation and unable to make progress on this issue the past week.

_Next Steps:_
Write test scripts that are started via Github Actions using Act/Nektos or via node.js.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
08/09/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
I have not had the time to work on the issue since the last update.

_Next Steps:_
Write test scripts that are started via Github Actions using Act/Nektos or via node.js.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
08/30/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by MarkWiltberger** [comment]
**Progress:**
I re-based my local development branch from the hackforla website GitHub repository. I looked at the git diffs for the new directory change, file location changes, filename changes, and code changes that have been made for other issues for the files under examination for this issue to see whether the changes would affect the testing of this issue. The code changes will not substantially affect the testing of the code.

_Next Steps:_
Next action is to install act/nektos locally in dev with new codebase and create a local e2e test for the file triggered by GitHub actions for the issue. For the issue, work with a git branch from the newly rebased code.

**Blockers:**
None.
**Availability:**
Daily 10am-6pm ET (7am - 3pm PT)
**ETA:**
09/13/2023 Wed 6pm ET (3pm PT)
**Pictures (optional):**
None.

**Comment by LRenDO** [comment]
Hi @MarkWiltberger! I was about to ask for an update, but I believe this issue may have been resolved by work on other issues, tagging in @t-will-gillis to verify. Regardless, I believe these files have been moved and refactored a bit and the links no longer work. @t-will-gillis, I think we both have worked in this GHA for a couple different issues. I worked on #4839 for updates getting applied when PRs are closed and I believe you worked on version updates as well as a refactor. As I can't see the errors referred to the workflow anymore, do you have any insight on this?

**Comment by t-will-gillis** [comment]
Hi @LRenDO - Yes, you are correct, this GHA has been refactored and updated many times so the original error may or may not be applicable anymore. 

@MarkWiltberger are you still working on this issue? From your previous comments it seems that you were working on a plan or outline automating the testing of the GHAs? That would be very valuable, and it seems the testing plan/process could be an epic broken into one or more large issues.

**Comment by MarkWiltberger** [comment]
Hello @t-will-gillis and @LRenDO ,

Yes, I am still working on this issue. @LRenDO , I have looked at the file changes and related issues that you mention. Most notably, these include changing of file names and locations. There have been some code changes as well.

@t-will-gillis , I agree the original error may or may not be applicable any more. I determined that in order to reproduce the error, a very helpful method would be to write meaningful automated tests.

I have been working on the method of how to create useful tests for a GitHub Action which calls a Node.js script.

I am thinking the next step for this issue is for me to finish creating a useful method of testing the particular GitHub Action for this issue, and see if it can be used to resolve this issue.

Then a separate issue could be created to document the testing method. Then the method could be applied to existing issues and applied to new issues, some of which could be grouped into an epic, if suitable.

**Progress:**
See above.

_Next Steps:_
See above.

**Blockers:**
None.

**Availability:**
Weekdays: Mornings 9am-11am ET (6am-8am PT)
Tue & Wed: Afternoons: 1pm-6pm (10am-3pm PT)

**ETA:**
10/25/2023 Wed 6pm ET (3pm PT)

**Pictures (optional):**
None.
